File: src/utils/debug.ts
const DEBUG_ENABLED =
  process.env.DEBUG_DOCMOST_MCP === "1" || process.env.DEBUG_DOCMOST_MCP === "true";

export function debugLog(scope: string, message: string, details?: unknown): void {
  if (!DEBUG_ENABLED) {
    return;
  }

  const timestamp = new Date().toISOString();
  const prefix = `[docmost-local-mcp][${timestamp}][${scope}]`;

  if (details === undefined) {
    process.stderr.write(`${prefix} ${message}\n`);
    return;
  }

  process.stderr.write(`${prefix} ${message} ${safeStringify(details)}\n`);
}

function safeStringify(value: unknown): string {
  try {
    return JSON.stringify(value);
  } catch {
    return String(value);
  }
}

File: src/debug.rs
use chrono::Utc;
use serde::Serialize;

pub fn debug_log<T>(scope: &str, message: &str, details: Option<&T>)
where
    T: Serialize + ?Sized,
{
    if !debug_enabled() {
        return;
    }

    let timestamp = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
    let prefix = format!("[docmost-local-mcp][{timestamp}][{scope}]");

    match details {
        Some(details) => {
            let serialized =
                serde_json::to_string(details).unwrap_or_else(|_| "<unserializable>".to_string());
            eprintln!("{prefix} {message} {serialized}");
        }
        None => eprintln!("{prefix} {message}"),
    }
}

pub fn debug_enabled() -> bool {
    matches!(
        std::env::var("DEBUG_DOCMOST_MCP").ok().as_deref(),
        Some("1") | Some("true")
    )
}
